soc/silabs/silabs_siwx917/siwg917/nwp_init.c

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright (c) 2024 Silicon Laboratories Inc.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/kernel.h>
+
+#include "sl_wifi.h"
+#include "sl_wifi_callback_framework.h"
+#if IS_ENABLED(CONFIG_BT_SIWX917)
+#include "rsi_ble_common_config.h"
+#endif
+
+/* TODO SILABS ZEPHYR: Check how to add constants to the vector table */
+#define NWP_ADD_TO_VECTOR_TABLE(irq, flags, expression, arg, name)                                 \
+	static Z_DECL_ALIGN(struct _isr_list) Z_GENERIC_SECTION(.intList)                          \
+		__used __isr_##name = {irq, flags, expression, arg}
+
+#if IS_ENABLED(CONFIG_WIFI_SIWX917) && IS_ENABLED(CONFIG_BT_SIWX917)
+#define NWP_INIT_COEX_MODE SL_SI91X_WLAN_BLE_MODE
+#elif IS_ENABLED(CONFIG_WIFI_SIWX917)
+#define NWP_INIT_COEX_MODE SL_SI91X_WLAN_ONLY_MODE
+#elif IS_ENABLED(CONFIG_BT_SIWX917)
+#define NWP_INIT_COEX_MODE SL_SI91X_BLE_MODE
+#else
+#error "Not supported configuration!"
+#endif
+
+static int silabs_siwx917_nwp_init(void)
+{
+	sl_wifi_device_configuration_t network_config = {
+		.boot_option = LOAD_NWP_FW,
+		.mac_address = NULL,
+		.band = SL_SI91X_WIFI_BAND_2_4GHZ,
+		.region_code = DEFAULT_REGION,
+		.boot_config = {
+			.oper_mode = SL_SI91X_CLIENT_MODE,
+			.coex_mode = NWP_INIT_COEX_MODE,
+			.feature_bit_map = SL_SI91X_FEAT_SECURITY_OPEN | SL_SI91X_FEAT_WPS_DISABLE,
+			.tcp_ip_feature_bit_map = SL_SI91X_TCP_IP_FEAT_EXTENSION_VALID,
+			.ext_tcp_ip_feature_bit_map = SL_SI91X_CONFIG_FEAT_EXTENSION_VALID,
+			.config_feature_bit_map = SL_SI91X_ENABLE_ENHANCED_MAX_PSP,
+			.custom_feature_bit_map = SL_SI91X_CUSTOM_FEAT_EXTENSION_VALID,
+			.ext_custom_feature_bit_map =
+				MEMORY_CONFIG | SL_SI91X_EXT_FEAT_XTAL_CLK |
+				SL_SI91X_EXT_FEAT_FRONT_END_SWITCH_PINS_ULP_GPIO_4_5_0,
+		}};
+	sl_si91x_boot_configuration_t *cfg = &network_config.boot_config;
+
+#if IS_ENABLED(CONFIG_WIFI_SIWX917)
+	cfg->tcp_ip_feature_bit_map |=
+#if IS_ENABLED(CONFIG_NET_IPV6)
+		SL_SI91X_TCP_IP_FEAT_DHCPV6_CLIENT | SL_SI91X_TCP_IP_FEAT_IPV6 |
+#endif
+		SL_SI91X_TCP_IP_FEAT_DHCPV4_CLIENT | SL_SI91X_TCP_IP_FEAT_DNS_CLIENT |
+		SL_SI91X_TCP_IP_FEAT_SSL | SL_SI91X_TCP_IP_FEAT_MDNSD | SL_SI91X_TCP_IP_FEAT_ICMP;
+	cfg->ext_tcp_ip_feature_bit_map |=
+		SL_SI91X_EXT_TCP_IP_WINDOW_SCALING | SL_SI91X_EXT_TCP_IP_TOTAL_SELECTS(10);
+#endif
```

Comment:
Here you can use `IS_ENABLED()`:
```
    if (IS_ENABLED(CONFIG_WIFI_SIWX917)) {
        cfg->ext_tcp_ip_feature_bit_map |= SL_SI91X_EXT_TCP_IP_WINDOW_SCALING;
        cfg->ext_tcp_ip_feature_bit_map |= SL_SI91X_EXT_TCP_IP_TOTAL_SELECTS(10);
        cfg->tcp_ip_feature_bit_map |= SL_SI91X_TCP_IP_FEAT_DHCPV4_CLIENT;
        cfg->tcp_ip_feature_bit_map |= SL_SI91X_TCP_IP_FEAT_DNS_CLIENT;
        cfg->tcp_ip_feature_bit_map |= SL_SI91X_TCP_IP_FEAT_MDNSD;
        cfg->tcp_ip_feature_bit_map |= SL_SI91X_TCP_IP_FEAT_ICMP;
        cfg->tcp_ip_feature_bit_map |= SL_SI91X_TCP_IP_FEAT_SSL;
    }
    if (IS_ENABLED(CONFIG_WIFI_SIWX917) && IS_ENABLED(CONFIG_NET_IPV6)) {
        cfg->tcp_ip_feature_bit_map |= SL_SI91X_TCP_IP_FEAT_DHCPV6_CLIENT;
        cfg->tcp_ip_feature_bit_map |= SL_SI91X_TCP_IP_FEAT_IPV6;
    }
```

Reply:
I know I could but I would not like to.
BT needs compile time selection because some macros are not defined if the feature is not selected (--> build error). This could happen with the WiFi as well in the future so I think it's better to keep as is.

Reply:
I assume we can fix these issues if we include `rsi_ble_common_config.h` unconditionally, that's it? If yes, I think it would make sense to do it.